include/glow/Partitioner/Partitioner.h

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_PARTITIONER_PARTITIONER_H
+#define GLOW_PARTITIONER_PARTITIONER_H
+
+#include "glow/Graph/Context.h"
+#include "glow/Graph/Graph.h"
+
+#include <map>
+#include <set>
+#include <string>
+
+namespace glow {
+
+using ModuleList = std::list<Module *>;
+
+struct DependencyDAG {
+  /// List of modules after partitioning. Each module will be assigned to run a
+  /// device by provioner.
+  ModuleList modules_;
+  /// The dependencies of each module.
+  Map modulesDAG_;
+  /// The root of this DAG.
+  Modules *root;
+  /// The number of nodes in the DAG.
+  int k;
+};
+
+/// Here we assume all the devices are identical.
+struct DeviceInfo {
+  /// Number of the devices.
+  int num;
+  /// Memory constrains of each device. unit GB.
+  int memSize;
+  /// TO DO... communication constrins and so on
+}
+
+/// Partition
+class Partitioner {
+
+  /// Dependency map type.
+  using Map = llvm::DenseMap<Module *, ModuleList>;
+
+  /// The module need to be decomposed.
+  Module *module_;
+
+  /// The cost model related to device.
+  DeviceInfo *deviceInfo_;
+
+  /// The result of module partitioning.
+  DependencyDAG dependencyDAG_;
+
+public:
+  /// \p parent is the module which contains the functions need to be divided.
+  /// Here we assume that all the functions in one module belong to a same
+  /// "Function Family", that is, without considerting the "dynamic stuff" (i.e.
+  /// batch size, input/output shape of each op), all the functions are
+  /// identical. The required memory and computation cost for each op can be
+  /// found in Module. The \p devices provides the cost model related to
+  /// devices.
+  Partitioner(const Module &parent, const DeviceInfo &devices);
+
+  /// Divide the module into minimal k sub-modules. The number of k is the
+  /// number of , where k is number of modules in the return value. k should be
+  /// less or equal to the number of devices. If we can't find a proper k
+  /// (though highly impossile I think. What I can think of is the size of some
+  /// constant > the momeries of all devices), I think anyway the partitioner
+  /// will return the result, and provisioner will check if k is proper?
+  DependencyDAG Partition();
```

Comment:
What's the ownership model of the DependencyDAG?  Will this do a copy?

Reply:
Yes, it will do a copy. The DependencyDAG will be saved in host manager.

Reply:
If it's going to be stored in the HostManager can it be moved out? I think avoiding copies of the module throughout the stack is going to make our lives much easier in the future.

Reply:
I think it is dependent on how HM works @gcatron.  From my point of view, Partitioner just provides the DependencyDAG to HM, and don't need to keep the result.

Reply:
The HM expects the DependencyDAG to be returned by the Partitioner. The DAG is then passed into the Provisioner and discarded at the end of the initGraph call. I think the Partitioner should be stateless.